File: java/trees/binary_tree_from_preorder_inorder.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    private HashMap<Integer, Integer> hash;
    private int[] gPreorder;
    private int[] gInorder;
    private int len;

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        len = preorder.length;
        if (len == 0) {
            return null;
        }
        hash = new HashMap<>();
        gPreorder = preorder;
        gInorder = inorder;
        for (int i = 0; i < inorder.length; i++) {
            hash.put(inorder[i], i);
        }

        return buildTreeUtil(0, preorder.length - 1, 0, inorder.length - 1);
    }

    public TreeNode buildTreeUtil(int start, int end, int startInorder, int endInorder){
        TreeNode head = new TreeNode(gPreorder[start]);

        if (start == end)
            return head;

        int index = hash.get(head.val);

        int leftStart = start + 1;
        int leftEnd = start + (index - startInorder);
        int rightStart = leftEnd + 1;
        int rightEnd = end;

        if (index != startInorder)
            head.left = buildTreeUtil(leftStart, leftEnd, startInorder, index - 1);
        
        if (index != endInorder)
            head.right = buildTreeUtil(rightStart, rightEnd, index + 1, endInorder);

        return head;
    }
}

File: python/trees/binaryTreeFromPreorderInorder.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        hash = {}
        for index in range(len(inorder)):
            hash[inorder[index]] = index
        def util(start: int, end: int, startInorder: int, endInorder: int):
            root = TreeNode(preorder[start])
            if start == end:
                return root
            
            index = hash[root.val]
			
            if index != startInorder:
                root.left = util(start+1, start+(index-startInorder), startInorder, index-1)
            if index != endInorder:
                root.right = util(start+(index-startInorder)+1, end, index+1, endInorder)
            return root
        return util(0, len(preorder)-1, 0, len(inorder)-1)
